File: code/ch13/13.3.1.best_time_to_buy_and_sell_stock_with_fee.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        if len(prices) <= 1:
            return 0
        # buy = [0 for i in range(len(prices))]
        # sell = [0 for i in range(len(prices))]
        buy, sell = [0] * len(prices), [0] * len(prices)
        buy[0] = -prices[0]
        for i in range(1, len(prices)):
            buy[i] = max(buy[i - 1], sell[i - 1] - prices[i])
            sell[i] = max(buy[i - 1] + prices[i] - fee, sell[i - 1])
        return sell[-1]

File: code/ch13/13.3.1.best_time_to_buy_and_sell_stock_with_fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        if (prices.size() <= 1)
            return 0;
        
	int res = 0;
        int *buy  = new int[prices.size() + 1]();
        int *sell = new int[prices.size() + 1]();
        
        buy[0] = -prices[0];
        
        for (int i = 1; i < prices.size(); i++) {
            buy[i]  = max(buy[i - 1], sell[i - 1] - prices[i]);
            sell[i] = max(buy[i - 1] + prices[i] - fee, sell[i - 1]);
        }

	res = sell[prices.size() - 1];
	delete [] buy;
	delete [] sell;
        
        return res;
    }
};
